pgjdbc/src/test/java/org/postgresql/test/util/PasswordUtilTest.java

```diff
@@ -0,0 +1,137 @@
+  }
+
+  private boolean tryUserPassword(String user, String password) {
+    Properties props = new Properties();
+    props.setProperty("user", user);
+    props.setProperty("password", password);
+    try (Connection conn = TestUtil.openDB(props)) {
+      return true;
+    } catch (SQLException e) {
+      return false;
+    }
+  }
+
+  private void testUserPassword(@Nullable String encryptionType, String username, String password,
+      String encodedPassword) throws SQLException {
+    String escapedUsername = Utils.escapeIdentifier(null, username).toString();
+
+    try (Connection superConn = TestUtil.openPrivilegedDB()) {
+      TestUtil.execute(superConn, "CREATE USER " //
+          + escapedUsername //
+          + " WITH PASSWORD '" + encodedPassword + "'");
+
+      String shadowPass = TestUtil.queryForString(superConn, //
+          "SELECT passwd FROM pg_shadow WHERE usename = ?", username);
+      Assert.assertEquals("pg_shadow value of password must match encoded", shadowPass, encodedPassword);
+
+      // We should be able to log in using our new user:
+      Assert.assertTrue("User should be able to authenticate using correct password",
+          tryUserPassword(username, password));
+      // We also check that we cannot log in with the wrong password to ensure that
+      // the server is not simply trusting everything
+      Assert.assertTrue("User should not be able to authenticate using incorrect password",
+          !tryUserPassword(username, "Bad Password:" + password));
+
+      String newPassword = "mySecretNewPassword" + randomSuffix();
+      PGConnection pgConn = superConn.unwrap(PGConnection.class);
+      pgConn.alterUserPassword(username, newPassword.toCharArray(), encryptionType);
+
+      // We should be able to log in using our new password
+      Assert.assertTrue("User should be able to authenticate using correct password",
+          tryUserPassword(username, newPassword));
+      // We also check that we cannot log in with the wrong password to ensure that
+      // the server is not simply trusting everything
+      Assert.assertTrue("User should not be able to authenticate using incorrect password",
+          !tryUserPassword(username, "Bad Password:" + newPassword));
+    } finally {
+      try (Connection superConn = TestUtil.openPrivilegedDB()) {
+        TestUtil.execute(superConn, "DROP USER " + escapedUsername);
+      } catch (Exception ignore) { }
+    }
+  }
+
+  private void testUserPassword(@Nullable String encryptionType, String username, String password) throws SQLException {
+    String encodedPassword = PasswordUtil.encodePassword(username, password.toCharArray(), encryptionType);
+    testUserPassword(encryptionType, username, password, encodedPassword);
+  }
+
+  private void testUserPassword(@Nullable String encryptionType) throws SQLException {
+    String username = "test_password_" + randomSuffix();
+    String password = "t0pSecret" + randomSuffix();
```

Comment:
Can you please add tests with passwords that include quotes, spaces, and the tests for non-encrypted password change?

Reply:
What's "non-encrypted password change"?

Reply:
Added the quotes etc but not sure what you meant by the last one.

Reply:
`encodePassword` supports `on`, `off`, `md5`, `scram-sha-256`. Can you please make sure you test all of them?

Reply:
Ah, you mean the `on` and `off` that are aliases for `md5`. Yes let's add those.